File: tests/recommendationservice/recommendation_service_stub.py
import os

from client import marshal_request, unmarshal_response, send_request, DEFAULT_TIMEOUT, \
    RECOMMENDATION_SERVICE, LIST_RECOMMENDATIONS_RPC
from logger import getJSONLogger

logger = getJSONLogger('recommendationservice-stub')


def list_recommendations(request, headers=None):
    bin_req = marshal_request(request)
    resp_body, headers = send_request(addr, RECOMMENDATION_SERVICE, LIST_RECOMMENDATIONS_RPC, bin_req, headers, timeout)
    return unmarshal_response(resp_body, headers, LIST_RECOMMENDATIONS_RPC)


addr = os.getenv("RECOMMENDATION_SERVICE_ADDR")
if addr is None:
    logger.error("RECOMMENDATION_SERVICE_ADDR environment variable not set")
    raise EnvironmentError("RECOMMENDATION_SERVICE_ADDR environment variable not set")

t = os.getenv("RECOMMENDATION_SERVICE_TIMEOUT")
if t is None:
    timeout = DEFAULT_TIMEOUT
else:
    try:
        timeout = int(t)
        if timeout <= 0:
            timeout = DEFAULT_TIMEOUT
    except ValueError:
        timeout = DEFAULT_TIMEOUT

File: src/frontend/client/recommendation_service_stub.go
package client

import (
    "log"
    "net/http"
    "os"
    "strconv"

    pb "main/genproto"
)

var (
    recommendationServiceAddr    *string
    recommendationServiceTimeout *int
)

const (
    recommendationService  = "recommendation-service"
    listRecommendationsRPC = "list-recommendations"
)

// ListRecommendations represents the RecommendationService/ListRecommendations RPC.
// context can be sent as custom headers.
func ListRecommendations(request *pb.ListRecommendationsRequest, header *http.Header) (*pb.ListRecommendationsResponse, error) {
    binReq, err := marshalRequest(request)
    if err != nil {
        return nil, err
    }

    respBody, header, err := sendRequest(*recommendationServiceAddr, recommendationService, listRecommendationsRPC, &binReq, header, *recommendationServiceTimeout)
    if err != nil {
        return nil, err
    }

    msg, err := unmarshalResponse(respBody, header, listRecommendationsRPC)
    if err != nil {
        return nil, err
    }

    return (*msg).(*pb.ListRecommendationsResponse), nil
}

// init loads the address and timeout variables.
func init() {
    a, ok := os.LookupEnv("RECOMMENDATION_SERVICE_ADDR")
    if !ok {
        log.Fatal("RECOMMENDATION_SERVICE_ADDR environment variable not set")
    }
    recommendationServiceAddr = &a

    t, ok := os.LookupEnv("RECOMMENDATION_SERVICE_TIMEOUT")
    if !ok {
        t := defaultTimeout
        recommendationServiceTimeout = &t
    } else {
        if t, err := strconv.Atoi(t); err != nil || t <= 0 {
            t = defaultTimeout
            recommendationServiceTimeout = &t
        } else {
            recommendationServiceTimeout = &t
        }
    }
}
